### ISSUE TITLE
Error 403 when music last played on a client (desktop or mobile) that has since been closed

### ISSUE THREAD

**Issue by Hannnes1** [fondberg/spotcast#490]
### The problem

When trying to run Spotcast after playing music on my phone or desktop and then closing the app, I get the following error. Tested on both an Android phone and Linux desktop.

```
Failed to perform the action script/new_script. http status: 403, code:-1 - https://api.spotify.com/v1/me/player/repeat?state=off&device_id=906527f2bf28cf7763d975bde7212649c5ede913: Player command failed: Restriction violated, reason: UNKNOWN
```

Steps to reproduce:

1. Open Spotify on a PC or mobile device
2. Play some music on that device
3. Pause the music
4. Close the app
5. Run Spotcast script in Home Assistant
6. See error

One thing to note is that the issue persists even if the app is reopened. Unless you start playing music again; then Spotcast works as long as the app isn't closed again.

## System Information

version | core-2024.11.3
-- | --
installation_type | Home Assistant OS
dev | false
hassio | true
docker | true
user | root
virtualenv | false
python_version | 3.12.4
os_name | Linux
os_version | 6.6.54-haos
arch | x86_64
timezone | Europe/Stockholm
config_dir | /config

<details><summary>Home Assistant Community Store</summary>

GitHub API | ok
-- | --
GitHub Content | ok
GitHub Web | ok
HACS Data | ok
GitHub API Calls Remaining | 4985
Installed Version | 2.0.1
Stage | running
Available Repositories | 1467
Downloaded Repositories | 4

</details>

<details><summary>Home Assistant Cloud</summary>

logged_in | false
-- | --
can_reach_cert_server | ok
can_reach_cloud_auth | ok
can_reach_cloud | ok

</details>

<details><summary>Home Assistant Supervisor</summary>

host_os | Home Assistant OS 13.2
-- | --
update_channel | stable
supervisor_version | supervisor-2024.11.4
agent_version | 1.6.0
docker_version | 27.2.0
disk_total | 30.8 GB
disk_used | 6.7 GB
healthy | true
supported | true
host_connectivity | true
supervisor_connectivity | true
ntp_synchronized | true
virtualization | kvm
board | ova
supervisor_api | ok
version_api | ok
installed_addons | Studio Code Server (5.17.3), Terminal & SSH (9.15.0), Mosquitto broker (6.4.1), Home Assistant Google Drive Backup (0.112.1)

</details>

<details><summary>Dashboards</summary>

dashboards | 3
-- | --
resources | 1
views | 1
mode | storage

</details>

<details><summary>Recorder</summary>

oldest_recorder_run | November 25, 2024 at 17:33
-- | --
current_recorder_run | November 25, 2024 at 18:40
estimated_db_size | 31.32 MiB
database_engine | sqlite
database_version | 3.45.3

</details>

<details><summary>Spotify</summary>

api_endpoint_reachable | ok
-- | --

</details>

### Spotcast Version

v4.0.0

### What type of installation are you running?

Home Assistant OS

### Diagnostics information

[home-assistant_spotcast_2024-11-28T19-46-15.540Z.log](https://github.com/user-attachments/files/17951940/home-assistant_spotcast_2024-11-28T19-46-15.540Z.log)

### Example YAML snippet

```yaml
Any script or automation should reproduce the issue, but here is the script I used:

alias: New Script
sequence:
  - action: spotcast.start
    data:
      device_name: Stereon
```

**Comment by fcusson** [comment]
Hi @Hannnes1,

I'm not quite sure that I understand. Are you trying to transfer the playback from your Spotify Desktop App, after you closed the app? If that's the case, then there is no active playback. Spotcast can't transfer anything.

Please confirm if I understood your request

**Comment by Hannnes1** [comment]
> Hi @Hannnes1,
> 
> I'm not quite sure that I understand. Are you trying to transfer the playback from your Spotify Desktop App, after you closed the app? If that's the case, then there is no active playback. Spotcast can't transfer anything.
> 
> Please confirm if I understood your request

Maybe? Probably. I never thought about it as transferring anything. I just assumed it would keep playing whatever I listened to last.

But when trying the Spotify API now I notice that `GET https://api.spotify.com/v1/me/player` returns an empty response after the client is closed. So I realize there is nothing you can do about that. 

A follow up question based on that though:
Would it be possible to fall back to playing your liked songs or a playlist if there is no active playback?

**Comment by fcusson** [comment]
@Hannnes1,

actually, I'm working on a way to save the last known playback and use that to recreate it on request when there is no active playback. That's something I'm planning with the current beta version of Spotcast. This might become available in a not too distant future.

Still I want to stress, I don't know If I'll be able to perfectly recreate the playback state, so there might be some drawback to it, we'll see how gracious the Spotify API is

**Comment by Hannnes1** [comment]
Sounds promising, thanks!

**Comment by RobertBuhren** [comment]
I'm experiencing the same 403 error. However, I'm a but confused by the comment that transferring playback is not supported. I used to use this snipped (as shown in the examples) to continue to play whatever I played last on my account on a chromecast device:

```
- service: spotcast.start
  data:
    device_name: "Speaker kitchen"
    force_playback: true
```

Is this not supported anymore?

**Comment by fcusson** [comment]
@RobertBuhren its only supported as long as the playback_state still exist in the Spotify API. When starting playback without a context uri, the API just restart the playback from any known states, but there are cases where the playback state is no longer reported. This is when the 403 error happens.

`V5` is using exactly the same implementation for transfering playback as `V3-V4` were. The only difference is that they are in 2 explicit services now (And in reality, when you call `spotcast.trasnsferPLayback` is just [call `spotcast.play_media`](https://github.com/fondberg/spotcast/blob/e938a879e70bbdc11b2c24c07c4572cf09c754d7/custom_components/spotcast/services/transfer_playback.py#L69) in the background).

Finally, I implemented a way to resume long gone playback for albums and playlists in the latest build. I'm working on the implementation for podcast (which is going to be a bit more complex)

*TL/DR:* `V4` and `V5` are doing the same thing in the background. If anything changed in the behavior, it might be Spotify who changed how long inactive playback are being reported through the API.

**Comment by RobertBuhren** [comment]
Ahh too bad. It used to work pretty well. I encountered the same issue with spotifyplus: https://github.com/thlucas1/homeassistantcomponent_spotifyplus/issues/66

I've used the script mentioned above successfully for quite some time. My guess would be that something in the spotify backend changed so that the playback_state is gone

**Comment by fcusson** [comment]
Build 16 releaed with ability to start playback from last known playback when no playback reported by Spotify

**Comment by Hannnes1** [comment]
Just tried it out; seems to work great. Thanks!
